### PR TITLE
Feature/pca for index

### PR DESCRIPTION
**Pull request by CarolinaVelasco**
This RMarkdown script does the following: 
- Randomly samples 20% of the original data
- Selects variables of interest for housing quality index
- Discards variables with more than 25% missing and 25% of "not applicable", "refuse to respond", "don't know", etc. 
- Performs PCA and builds index
- Selects demographic variables of interest

The final output is an R file that contains the demographic variables of interest, the component scores, and the standardized index.

### CODE REVIEW

**Comment by lindechen** [comment]
Can someone upload either the full dataset or the 20% sampled dataset to the Data folder in Github? I see that @CarolinaVelasco was reading the full dataset from her Dropbox and the 20% dataset is not yet pulled, so I want to make sure I'm running code on the same data as you guys.

**Comment by lindechen** [comment]
Actually, nvm - I found the original data. ^

**Comment by lindechen** [comment]
Question for discussion:

Is it significant (do we care) that for the purposes of future analysis (logistic regression and XGBoost) that 1) the distribution of the HQ index is bimodal, and 2) there are outliers in the upper end?

**Comment by lindechen** [comment]
> Sure! There is not a standard and objective way to decide the number of principal components. I have read in papers different ways to determine the number of components: (1) eyeballing the scree plot and finding the elbow, (2) keep all components whose PVE is greater than 1, and/or (3) choose all the components that show interesting patterns in the data. I found interesting patters (i.e. the component was mostly correlated with variables that represented a specific housing characteristic) up to component 6. We can definitely choose less / more components, it is up to us.

**Review by lindechen — APPROVED** [review]
Looks good to me

**Review by jcochanc — CHANGES_REQUESTED** [review]
Hi @CarolinaVelasco,

Overall I think this is a great job. As @lindechen mentioned, very clear explanations and informative plots. There are a few changes I think should happen before we can use some of the data here. For a step by step, see my comments but I will try and summarize here.

* In a few cases, data that is written as saveRDS is not found in the most current branch; this may be an issue on my end or when that was run, the push was never committed. Those files should be saved and available for use downstream. Again, apologies if this is something wrong on my end.

* You mentioned there are no objective ways for scree selection but I have read about two; would you care to elaborate why those are not being considered? Balancing the plot and the one that follows, I would agree six is the way to go. Just curious.

* Finally, why is the index out of 10? Were there 10 factors we used? If we only used 6 loadings, shouldn't the index be out of 6? My concern is that if 10 is interpreted as the best housing possible, we only have 0.004187953% of the data with scores above 4; essentially none of these housing units are good.

**Comment by CarolinaVelasco** [comment]
> What about using the scree test or the Cattell–Nelson–Gorsuch scree test?

I have not heard about these tests. I'll look into them. Have you done them before / do you have any resources?

**Comment by CarolinaVelasco** [comment]
> @CarolinaVelasco When I run this, I get this warning: "Missing values are imputed by the mean of the variable: you should use the imputePCA function of the missMDA package"; mean value imputation has the downside of reducing the variation in the data. Is there another option?

@jcochanc I noticed the message and tried using the missMDA package. However, I got some errors and did not have enough time to go through them. If I have time, I'll explore this.

**Comment by CarolinaVelasco** [comment]
> Hi @CarolinaVelasco,
> 
> Overall I think this is a great job. As @lindechen mentioned, very clear explanations and informative plots. There are a few changes I think should happen before we can use some of the data here. For a step by step, see my comments but I will try and summarize here.
> 
> * In a few cases, data that is written as saveRDS is not found in the most current branch; this may be an issue on my end or when that was run, the push was never committed. Those files should be saved and available for use downstream. Again, apologies if this is something wrong on my end.
> * You mentioned there are no objective ways for scree selection but I have read about two; would you care to elaborate why those are not being considered? Balancing the plot and the one that follows, I would agree six is the way to go. Just curious.
> * Finally, why is the index out of 10? Were there 10 factors we used? If we only used 6 loadings, shouldn't the index be out of 6? My concern is that if 10 is interpreted as the best housing possible, we only have 0.004187953% of the data with scores above 4; essentially none of these housing units are good.

@jcochanc  We can change the index to be out of any number (I arbitrarily chose 10). However, as expected, the distribution of the index will look the same regardless of the range we choose.  The 6 factors were used to build the index. Refer to the RMarkdown file for the specifics as to how it was computed. I was also concerned about the distribution of the index.  I would appreciate any suggestions / other ideas  to go about the creation of the index with the 6 components.

**Comment by jcochanc** [comment]
> @jcochanc We can change the index to be out of any number (I arbitrarily chose 10). However, as expected, the distribution of the index will look the same regardless of the range we choose. The 6 factors were used to build the index. Refer to the RMarkdown file for the specifics as to how it was computed. I was also concerned about the distribution of the index. I would appreciate any suggestions / other ideas to go about the creation of the index with the 6 components.

I guess I misunderstood the method. My understanding is that an index is not continuous but rather a sum of 1’s or 0’s. This index seems to be continuous - is that typical? 

If we kept it in this zero or one format, then for every factor loading, if a person had some percentage over for that factor loading, they would get a 1, 0 otherwise. Then the index would be out of 6 and integer valued.

**Comment by CarolinaVelasco** [comment]
> > @jcochanc We can change the index to be out of any number (I arbitrarily chose 10). However, as expected, the distribution of the index will look the same regardless of the range we choose. The 6 factors were used to build the index. Refer to the RMarkdown file for the specifics as to how it was computed. I was also concerned about the distribution of the index. I would appreciate any suggestions / other ideas to go about the creation of the index with the 6 components.
> 
> I guess I misunderstood the method. My understanding is that an index is not continuous but rather a sum of 1’s or 0’s. This index seems to be continuous - is that typical?
> 
> If we kept it in this zero or one format, then for every factor loading, if a person had some percentage over for that factor loading, they would get a 1, 0 otherwise. Then the index would be out of 6 and integer valued.

I have read in the literature indices that are continuous, especially those that measure socioeconomic status, but it really depends on the methodology being used to build the index. For the purpose of our research, the index could be either categorical or continuous. What is a common choice of threshold for the factor loadings to build the index in the zero or one format?

**Comment by CarolinaVelasco** [comment]
> Oh, but I see your point now @jcochanc. The data is not currently in the Data folder in this branch.

I just uploaded the data again, for some reason it is not there anymore. Please let me know if you have access to it now.

**Comment by CarolinaVelasco** [comment]
> > Oh, but I see your point now @jcochanc. The data is not currently in the Data folder in this branch.
> 
> I just uploaded the data again, for some reason it is not there anymore. Please let me know if you have access to it now.

Actually, I made a mistake right now and committed the upload under the wrong branch. I'll work on fixing it now, sorry!

**Comment by jcochanc** [comment]
> Question for discussion:
> 
> Is it significant (do we care) that for the purposes of future analysis (logistic regression and XGBoost) that 1) the distribution of the HQ index is bimodal, and 2) there are outliers in the upper end?

@lindechen I did not even consider that! Good catch. Maybe wherever we land, we can have this better distributed.

> > > @jcochanc We can change the index to be out of any number (I arbitrarily chose 10). However, as expected, the distribution of the index will look the same regardless of the range we choose. The 6 factors were used to build the index. Refer to the RMarkdown file for the specifics as to how it was computed. I was also concerned about the distribution of the index. I would appreciate any suggestions / other ideas to go about the creation of the index with the 6 components.
> > 
> > If we kept it in this zero or one format, then for every factor loading, if a person had some percentage over for that factor loading, they would get a 1, 0 otherwise. Then the index would be out of 6 and integer valued.
> 
> I have read in the literature indices that are continuous, especially those that measure socioeconomic status, but it really depends on the methodology being used to build the index. For the purpose of our research, the index could be either categorical or continuous. What is a common choice of threshold for the factor loadings to build the index in the zero or one format?

@CarolinaVelasco Awesome! I have not come across any but will defer to your expertise :) Do you happen to have a list of those references? I don't need them but it would be nice to cite those in the paper/slides. As for common thresholds, usually people have used above 0.75 for each item for it to be 1. For example ifelse(factor_loading_1 > 0.75, 1, 0).

**Comment by CarolinaVelasco** [comment]
> > > Oh, but I see your point now @jcochanc. The data is not currently in the Data folder in this branch.
> > 
> > 
> > I just uploaded the data again, for some reason it is not there anymore. Please let me know if you have access to it now.
> 
> Actually, I made a mistake right now and committed the upload under the wrong branch. I'll work on fixing it now, sorry!

Solved!

**Comment by CarolinaVelasco** [comment]
> > Question for discussion:
> > Is it significant (do we care) that for the purposes of future analysis (logistic regression and XGBoost) that 1) the distribution of the HQ index is bimodal, and 2) there are outliers in the upper end?
> 
> @lindechen I did not even consider that! Good catch. Maybe wherever we land, we can have this better distributed.
> 
> > > > @jcochanc We can change the index to be out of any number (I arbitrarily chose 10). However, as expected, the distribution of the index will look the same regardless of the range we choose. The 6 factors were used to build the index. Refer to the RMarkdown file for the specifics as to how it was computed. I was also concerned about the distribution of the index. I would appreciate any suggestions / other ideas to go about the creation of the index with the 6 components.
> > > 
> > > 
> > > If we kept it in this zero or one format, then for every factor loading, if a person had some percentage over for that factor loading, they would get a 1, 0 otherwise. Then the index would be out of 6 and integer valued.
> > 
> > 
> > I have read in the literature indices that are continuous, especially those that measure socioeconomic status, but it really depends on the methodology being used to build the index. For the purpose of our research, the index could be either categorical or continuous. What is a common choice of threshold for the factor loadings to build the index in the zero or one format?
> 
> @CarolinaVelasco Awesome! I have not come across any but will defer to your expertise :) Do you happen to have a list of those references? I don't need them but it would be nice to cite those in the paper/slides. As for common thresholds, usually people have used above 0.75 for each item for it to be 1. For example ifelse(factor_loading_1 > 0.75, 1, 0).

@jcochanc I do have some references, I'll compile them for the paper if we end up using the continuous index. I am a bit confused about the factor loading method using the threshold. From my understanding, when we do PCA we have component loadings for every variable and component scores for every individual. Therefore, I think the threshold should be chosen for the scores of each individual (i.e. we assign a 0 or 1 to each individual for every component.) The scores are continuous and I'm not sure how to pick a threshold for them. Is my understanding accurate?

**Comment by jcochanc** [comment]
> @jcochanc I do have some references, I'll compile them for the paper if we end up using the continuous index. I am a bit confused about the factor loading method using the threshold. From my understanding, when we do PCA we have component loadings for every variable and component scores for every individual. Therefore, I think the threshold should be chosen for the scores of each individual (i.e. we assign a 0 or 1 to each individual for every component.) The scores are continuous and I'm not sure how to pick a threshold for them. Is my understanding accurate?

A component has a minimum value and a maximum value relative to the variables included. Suppose that a component is made up of var1 and var to and var1 is ordinal with three values (1-3) and var2 is ordinal with 2 values (1-2), then at least you would have the 1 and at most you could have 5, then these values times their respective component values (the pieces estimated by PCA) would give you a true min and true max. We could then use the top quartile as a cutoff.  

If you see table 4, page 12, (https://unstats.un.org/unsd/hhsurveys/finalpublication/ch18fin3.pdf) the first two for component 1 are 0.24 and 0.39. If we assume they take the vlaues of var1 and var2, then the min is 0.24*1+0.39*1 while the max is 0.24*3+0.39*2. Then for everyone we could calculate the quartile belonging and score 1 for the upper and zero for belonging to the bottom three. Just an idea!

**Comment by CarolinaVelasco** [comment]
> > @jcochanc I do have some references, I'll compile them for the paper if we end up using the continuous index. I am a bit confused about the factor loading method using the threshold. From my understanding, when we do PCA we have component loadings for every variable and component scores for every individual. Therefore, I think the threshold should be chosen for the scores of each individual (i.e. we assign a 0 or 1 to each individual for every component.) The scores are continuous and I'm not sure how to pick a threshold for them. Is my understanding accurate?
> 
> A component has a minimum value and a maximum value relative to the variables included. Suppose that a component is made up of var1 and var to and var1 is ordinal with three values (1-3) and var2 is ordinal with 2 values (1-2), then at least you would have the 1 and at most you could have 5, then these values times their respective component values (the pieces estimated by PCA) would give you a true min and true max. We could then use the top quartile as a cutoff.
> 
> If you see table 4, page 12, (https://unstats.un.org/unsd/hhsurveys/finalpublication/ch18fin3.pdf) the first two for component 1 are 0.24 and 0.39. If we assume they take the vlaues of var1 and var2, then the min is 0.24_1+0.39_1 while the max is 0.24_3+0.39_2. Then for everyone we could calculate the quartile belonging and score 1 for the upper and zero for belonging to the bottom three. Just an idea!

@jcochanc I like this idea, I'll try it tonight. Do you happen to have sources that have used this method? I've never heard of it before and would like to learn more.

**Comment by CarolinaVelasco** [comment]
@lindechen @jcochanc I just created the alternative index following @jcochanc suggestion. I sent you a re-review.

**Review by jcochanc — APPROVED** [review]
@CarolinaVelasco this looks great! All the data is visible now in the folder and the new index looks good. I saw your comment about the same percent above 4; but it is better than having a huge gap from 4 continuous to 10 than 4 to 5. I hope that makes sense?

File: pca_for_index/pca_analysis.Rmd [hunk 1]
```diff
@@ -0,0 +1,439 @@
...
+comp5_score_binary <- ifelse(comp5_score > quantile(comp5_score, c(0.75), na.rm = TRUE), 1, 0)
+comp6_score_binary <- ifelse(comp6_score > quantile(comp6_score, c(0.75), na.rm = TRUE), 1, 0)
+
+# Create index as the sum of the binary scores. The index should have a range 0 - 6. However, similar to what was seen in the first housing quality index that was created, only 0.04036582% of the data has scores above 4. Additionally, this method of creating the index does not account for missing values. A total of 3089 observations have missing values and, hence, are not assigned an index.
+
+hq_index2 <- comp1_score_binary + comp2_score_binary + comp3_score_binary + comp4_score_binary + comp5_score_binary + comp6_score_binary
+
+summary(hq_index2)
+length(which(hq_index2 > 4))*100/nrow(df_quality_vars_lumped_reduced_num) # only 0.04% of the data has scores above 4
+hist(hq_index2)
```
Comment (lindechen) [inline pca_for_index/pca_analysis.Rmd:326]: If we include this plot in the paper or presentation, can we add breaks=5 to the histogram so it makes the x-axis look discrete? Right now we have some space in between each bars, which makes the index look continuous.

File: pca_for_index/pca_analysis.Rmd [hunk 2]
```diff
@@ -0,0 +1,439 @@
...
+hist(std_hq_index, breaks=100)
+```
+
+
+```{r}
+# Create dataset that contains component scores and standardized index
+df_index <- scores %>% mutate(housing_qual_index = std_hq_index)
+```
+
+# Alternative way to build the index
```
Comment (lindechen) [inline pca_for_index/pca_analysis.Rmd:292]: I like this alternative index a lot! I'd be happy to mention just this one in the presentation / paper.

**Review by lindechen — APPROVED** [review]
Looks great to me! Just one minor comment about the histogram plot :)

**Pull request merged**